// @cainc/game-common/src/js/game_states/MainMenu.js
import {
    ExitButton,
    gameBridge,
    HelpButton,
    StartButton,
    ToggleAudioButton
} from '../index.js';

const LAYOUT = {
    START_X: 497,
    START_Y: 421,

    EXIT_X: 1185,
    EXIT_Y: 10,

    HELP_X: 10,
    HELP_Y: 870,

    AUDIO_X: 1185,
    AUDIO_Y: 870,
};
Object.freeze(LAYOUT);

export default class MainMenu extends Phaser.State {
    init(helpScreens) {
        this._helpScreens = helpScreens;
    }

    create() {
        this.add.image(0, 0, 'main_menu_background');
        this.startButton = new StartButton(this, LAYOUT.START_X, LAYOUT.START_Y, 'start_button', this._helpScreens);
        if (gameBridge.shouldShowExitButton) {
            this.exitButton = new ExitButton(this, LAYOUT.EXIT_X, LAYOUT.EXIT_Y, 'exit_button');
        }
        this.helpButton = new HelpButton(this, LAYOUT.HELP_X, LAYOUT.HELP_Y, 'help_button', this._helpScreens);
        this.toggleAudioButton = new ToggleAudioButton(this, LAYOUT.AUDIO_X, LAYOUT.AUDIO_Y, 'toggle_audio');

        displayGame();
    }
}



//////////////////
// WEBPACK FOOTER
// ./node_modules/@cainc/game-common/src/js/game_states/MainMenu.js
// module id = 342
// module chunks = 0
