// @cainc/game-common/src/js/util/GameBridge.js
class InternalizedGameBridge {
    constructor() {
        if (parent.window.gameBridge) {
            _.forEach(parent.window.gameBridge, (field, name) => {
                this[name] = field;
            });
            parent.window.gameBridge = undefined;
        } else {
            this.start = () => {
                return new Promise((resolve) => {
                    resolve();
                });
            };
            this.close = _.noop;
            this.initialized = _.noop;
            this.sendScore = undefined;
            this.updateLoadProgress = _.noop;
            this.getUserData = undefined;
            this.saveUserData = undefined;
            this.info = {
                numberOfRetries: 1,
                studentId: 3141592653,
            };
        }
    }

    get usingRealBridge() {
        return this.sendScore !== undefined;
    }

    /**
     * We show the exit button when we are not in the diag
     */
    get shouldShowExitButton() {
        return (this.isWithinDiag !== true);
    }
}


export const gameBridge = new InternalizedGameBridge();



//////////////////
// WEBPACK FOOTER
// ./node_modules/@cainc/game-common/src/js/util/GameBridge.js
// module id = 51
// module chunks = 0
