// @cainc/game-common/src/js/components/ToggleAudioButton.js
export default class ToggleAudioButton {
    /**
     * @param  {State} gameState    the current state - used as a parent for the button
     * @param  {number} x           left coordinate of the button
     * @param  {number} y           top coordinate of the button
     * @param  {AssetOptions} asset see typedef in StartButton
     */
    constructor(gameState, x, y, asset) {
        const currentAssets =
            gameState.sound.mute ? {
                over: asset.offOver,
                out: asset.offOut,
                down: asset.offDown,
                up: asset.offUp,
            } :
            {
                over: asset.over,
                out: asset.out,
                down: asset.down,
                up: asset.up,
            };

        this._button = gameState.add.button(x, y, asset.atlas, () => {
            if (gameState.sound.mute) {
                // toggle audio on
                this._button.setFrames(asset.over, asset.out, asset.down, asset.up);
                gameState.sound.mute = false;
            } else {
                // toggle audio off
                this._button.setFrames(asset.offOver, asset.offOut, asset.offDown, asset.offUp);
                gameState.sound.mute = true;
            }
        }, null, currentAssets.over, currentAssets.out, currentAssets.down, currentAssets.up);
        this._button.setUpSound(asset.upSound);
    }

    get button() {
        return this._button;
    }
}



//////////////////
// WEBPACK FOOTER
// ./node_modules/@cainc/game-common/src/js/components/ToggleAudioButton.js
// module id = 340
// module chunks = 0
